### ISSUE TITLE
Show data protection warning when downloading files

### ISSUE THREAD

**Issue by dc2007git** [rcpch/national-paediatric-diabetes-audit#118]
More of a TODO than an issue - but logging here for reference:

<img width="474" alt="image" src="https://github.com/rcpch/national-paediatric-diabetes-audit/assets/65614251/abda9293-5d6c-4ada-b33b-9407df94d6dd">

The download button in the audit cohort view still requires functionality to be integrated. I imagine it should just redownload the same csv as was submitted for that instance.

**Comment by mbarton** [comment]
To match the security measures of the previous platform we should wrap any downloads in a password encrypted ZIP file. The old platform let you specify your own password but I would prefer we generate a one time password displayed after the download

**Comment by mbarton** [comment]
I think it's important to download *literally the same CSV file* as was uploaded - so implementing this should be less of an "export" option and more "download what the user uploaded to help debugging"

**Comment by mbarton** [comment]
Implemented in #257 

The old system puts any downloads in a password protected ZIP file but we haven't done that for people downloading the CSV files. It seems maybe overkill since they were likely to have submitted the files in the first place.

@AmaniKrayemRCPCH should we password protect the CSV when people download it?

**Comment by AmaniKrayemRCPCH** [comment]
Has something similar been implemented in E12? Do they require a password?

**Comment by mbarton** [comment]
I don't think we ever implemented it in E12, there was some discussion late last year https://github.com/rcpch/rcpch-audit-engine/issues/643

Do data uploaders need download access anyway since they had the spreadsheet unencrypted beforehand? If we limit the feature to the NPDA team then I would say we don't need it as we will already store and handle it appropriately.

**Comment by AmaniKrayemRCPCH** [comment]
Having looked at our DPIA, we state that any data downloaded by units is password protected, so we need to stick with that. 

It's hard to tell how often people are actually exporting their data - it's not in our user event log. Nor can I think of a use case. 
Those submitting via CSV will already have their data file, but those submitting via questionnaire won't. 

I can get some user perspective at the next dataset and methodology working group or project board meeting?

**Comment by mbarton** [comment]
Yeah this is specifically for downloading the CSV you uploaded, which I think was in the original requirements so the NPDA team could investigate claims that numbers were incorrect.

We haven't implemented data export for people submitting via questionnaire yet.

> I can get some user perspective at the next dataset and methodology working group or project board meeting?

That would be very useful thanks, we can use that to guide what we password protect and if we need to give people other than the NPDA team access to download the CSV files

**Comment by mbarton** [comment]
@AmaniKrayemRCPCH circling back to this issue now. I'm leaning towards not giving people the option to download the last uploaded CSV file? RCPCH audit team and admins will be able to but we wouldn't offer it to regular users.

**Comment by AmaniKrayemRCPCH** [comment]
I think that's fine!

If a unit is absolutely desperate for a download of their data, I can share it with them via SharePoint (with IG approval). 

Does this affect the data quality report download?

**Comment by mbarton** [comment]
Ah of course, our data quality report includes the original data and data with comments as separate sheets.

So the raw data is inherent to that feature and it's so useful I'd be worried about removing it, although obviously just the error, nhs_number and CSV row in a sheet is enough to troubleshoot errors.

**Comment by AmaniKrayemRCPCH** [comment]
I can look at updating our DPIA, if that's preferable to making downloads password protected?

**Comment by mbarton** [comment]
It’s not difficult to add but I worry it would make the data quality report download annoying.

If we do passwords it would be a one time thing: it shows you the password as you download the file and then you can’t see it again. I think that’s better than letting people pick their own passwords - that defeats the point of the feature as they will be reused.

Perhaps we add it to be in compliance with the DPIA and also look to update it at the same time?

**Comment by AmaniKrayemRCPCH** [comment]
I'm happy with that! 
Agreed that it could make the data quality report annoying, but I'm happy to add it in and then ask about it at UAT

**Comment by mbarton** [comment]
We discussed this in the NPDA catch up call today and naturally as soon as the call finished I had another idea.

The reason for encrypting the download is that it contains bulk PII and is ending up in an environment we don't control. But of course we also don't control the environment the user is uploading from, so there's no value in password protecting the download if done immediately after upload. It's only ending up where the data came from in the first place!

So perhaps a way forward is offer users the data quality report for download in the UI immediately after uploading. It would not be encrypted. If you then download the original data or data quality report at a later date, it is encrypted.

The technical implementation would be a "burn after reading" link for the unencrypted report, tied to the session the upload was performed in which expires a fixed amount of time after upload.

What do you think @AmaniKrayemRCPCH @eatyourpeas?

**Comment by AmaniKrayemRCPCH** [comment]
That makes sense to me, and would mean that, the majority of the time, users wouldn't need to use a password. 

I think we should get IG input on the final decision.

**Comment by mbarton** [comment]
I think it’s possible to share Excel files as View Only from Sharepoint, without the ability to download.

If anything that’s a stronger protection than password protecting the file - from a data management point of view it never leaves an environment we control.

We do try to avoid adding provider specific functionality to our audit platforms but you could imagine this would be useful as a generic function in the future - share an export out using Sharepoint, S3 presigned URLs etc

**Comment by AmaniKrayemRCPCH** [comment]
Like the SharePoint idea - the only issue that comes to mind is we've had trouble sharing files via sharepoint in the past with some NHS users

On E12, we sent out reports to each unit via SharePoint annual and, each year, a number of units would have trouble accessing it because they didn't have an NHS sharepoint account, or their trust's sharepoint would block them from accessing our sharepoint site somehow.

**Comment by mbarton** [comment]
Yeah if we do go down that Sharepoint route ideally it would be a link that works without logging in and is automatically revoked after a certain period. Looks like that's [not possible though](https://answers.microsoft.com/en-us/msoffice/forum/all/accessing-sharepoint-site-without-login-to/140d9977-c213-4fa4-8d7f-25d3d7253cf2) :(

**Comment by mbarton** [comment]
Agreed with the RCPCH Information Governance team that we will not password protect the data quality reports. We will show a message advising good data protection practice before the download (and downloading the original CSV file), something like:

> The data quality report contains your original data annotated with errors using our validation rules.
> It contains personal information and must be stored securely and deleted after use.

**Comment by dc2007git** [comment]
@mbarton do we want to log the data protection message before the user downloads a CSV of their already uploaded data? I saw an earlier comment you made where you pointed out that the user will already have uploaded the exact same file, so it might not be relevant?

**Comment by mbarton** [comment]
Ah yes good point, we should probably show it there too. I guess the message needs to be slightly different there too:

> This download contains your entire original data file.
> It contains personal information and must be stored securely and deleted after use.
